public/app/features/dashboard/components/DashboardPrompt/DashboardPrompt.tsx

```diff
@@ -0,0 +1,217 @@
+import React, { useEffect, useState } from 'react';
+import { Prompt } from 'react-router-dom';
+import { DashboardModel } from '../../state/DashboardModel';
+import { each, filter, find } from 'lodash';
+import angular from 'angular';
+import { UnsavedChangesModal } from '../SaveDashboard/UnsavedChangesModal';
+import * as H from 'history';
+import { SaveLibraryPanelModal } from 'app/features/library-panels/components/SaveLibraryPanelModal/SaveLibraryPanelModal';
+import { PanelModelWithLibraryPanel } from 'app/features/library-panels/types';
+import { useDispatch } from 'react-redux';
+import { discardPanelChanges } from '../PanelEditor/state/actions';
+import { DashboardSavedEvent } from 'app/types/events';
+
+export interface Props {
+  dashboard: DashboardModel;
+}
+
+interface State {
+  original: object | null;
+  originalPath?: string;
+  modal: PromptModal | null;
+  blockedLocation?: H.Location | null;
+}
+
+enum PromptModal {
+  UnsavedChangesModal,
+  SaveLibraryPanelModal,
+}
+
+export const DashboardPrompt = React.memo(({ dashboard }: Props) => {
+  const [state, setState] = useState<State>({ original: null, modal: null });
+  const dispatch = useDispatch();
+  const { original, originalPath, blockedLocation, modal } = state;
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => {
+      const originalPath = locationService.getLocation().pathname;
+      const original = dashboard.getSaveModelClone();
+
+      setState({ originalPath, original, modal: null });
+    }, 1000);
+
+    const savedEventUnsub = appEvents.subscribe(DashboardSavedEvent, () => {
+      const original = dashboard.getSaveModelClone();
+      const originalPath = locationService.getLocation().pathname;
+      setState({ originalPath, original, modal: null });
+    });
+
+    return () => {
+      clearTimeout(timeoutId);
+      savedEventUnsub.unsubscribe();
+    };
+  }, [dashboard]);
+
+  const onHistoryBlock = (location: H.Location) => {
+    const panelInEdit = dashboard.panelInEdit;
+    const search = new URLSearchParams(location.search);
+
+    // Are we leaving panel edit & library panel?
+    if (panelInEdit && panelInEdit.libraryPanel && panelInEdit.hasChanged && !search.has('editPanel')) {
+      setState({ ...state, modal: PromptModal.SaveLibraryPanelModal, blockedLocation: location });
+      return false;
+    }
+
+    // Are we still on the same dashboard?
+    if (originalPath === location.pathname || !original) {
+      return true;
+    }
+
+    if (ignoreChanges(dashboard, original)) {
+      return true;
+    }
+
+    if (!hasChanges(dashboard, original)) {
+      return true;
+    }
+
+    setState({ ...state, modal: PromptModal.UnsavedChangesModal, blockedLocation: location });
+    return false;
+  };
+
+  return (
+    <>
+      <Prompt when={true} message={onHistoryBlock} />
+      {modal === PromptModal.UnsavedChangesModal && (
+        <UnsavedChangesModal
+          dashboard={dashboard}
+          onSaveSuccess={() => {
```

Comment:
Would `useCallback` make sense here?

Reply:
no, DashboardPrompt is only re-rendered when dashboards change or on location change, so caching the callback would have no effect